File: day05/day05.go
package day05

import (
	"fmt"
	"strings"
)

type Command struct{}

func (cmd *Command) Execute(input string) (any, any) {
	return cmd.part1(input), cmd.part2(input)
}

func (cmd *Command) part1(input string) string {
	stacks := cmd.parseStacks(input)

	for _, line := range strings.Split(input, "\n") {
		if strings.HasPrefix(line, "move") {
			var toMove int
			var from int
			var to int
			fmt.Sscanf(line, "move %d from %d to %d", &toMove, &from, &to)

			for move := 0; move < toMove; move++ {
				stacks[to-1].Push(stacks[from-1].Pop())
			}
		}
	}

	result := ""
	for _, s := range stacks {
		result += string(s.Pop())
	}

	return result
}

func (cmd *Command) part2(input string) string {
	stacks := cmd.parseStacks(input)

	for _, line := range strings.Split(input, "\n") {
		if strings.HasPrefix(line, "move") {
			var toMove int
			var from int
			var to int
			fmt.Sscanf(line, "move %d from %d to %d", &toMove, &from, &to)

			stacks[to-1].Push(stacks[from-1].PopN(toMove)...)
		}
	}

	result := ""
	for _, s := range stacks {
		result += string(s.Pop())
	}

	return result
}

func (cmd *Command) parseStacks(input string) []Stack {
	stacks := []Stack{}

	parts := strings.Split(input, "\n\n")
	lines := strings.Split(parts[0], "\n")

	// Needed as we need to know the number of stacks
	size := len(strings.Fields(lines[len(lines)-1]))
	for i := 0; i < size; i++ {
		stacks = append(stacks, Stack{})
	}

	for _, line := range lines[:len(lines)-1] {
		for i, r := range line {
			if r != ' ' && r != '[' && r != ']' {
				stacks[i/4].AddToBottom(r)
			}
		}
	}

	return stacks
}

File: day05/day05.ts
const part1 = (parsedInput: string[][]): string => {
  const [input, moves] = parsedInput;

  const stacks: string[][] = [];

  for (let i = 0; i < 9; i++) {
    const stack: string[] = [];

    for (let j = input.length - 2; j >= 0; j--) {
      const char = input[j][i * 4 + 1]?.trim();
      if (char) {
        stack.push(char);
      }
    }

    stacks.push(stack);
  }

  moves.forEach((move) => {
    const [no, fr, to] = move
      .split(/(move | from | to )/g)
      .filter((l) => !/[\sa-z]/.test(l) && l.length > 0)
      .map(Number);

    for (let j = 0; j < no; j++) {
      const move = stacks[fr - 1].pop();
      stacks[to - 1].push(move!);
    }
  });

  return stacks.map((s) => s.pop()).join("");
};

const part2 = (parsedInput: string[][]): string => {
  const [input, moves] = parsedInput;

  const stacks: string[][] = [];

  for (let i = 0; i < new Number(input.at(-1)?.trim().at(-1)); i++) {
    const stack: string[] = [];

    for (let j = input.length - 2; j >= 0; j--) {
      const char = input[j][i * 4 + 1]?.trim();
      if (char) {
        stack.push(char);
      }
    }

    stacks.push(stack);
  }

  moves.forEach((move) => {
    const [no, fr, to] = move
      .split(/(move | from | to )/g)
      .filter((l) => !/[\sa-z]/.test(l) && l.length > 0)
      .map(Number);

    const moving: string[] = [];
    for (let j = 0; j < no; j++) {
      moving.push(stacks[fr - 1].pop()!);
    }

    moving.reverse().forEach((m) => stacks[to - 1].push(m));
  });

  return stacks.map((s) => s.pop()).join("");
};

export default function (rawInput: string): [string, string] {
  const parsedInput = rawInput
    .split("\n\n")
    .map((p) => p.split("\n"));

  return [part1(parsedInput), part2(parsedInput)];
}
